File: graph_neural_network/C++/Grok4_JSSP_GNN.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cmath>
#include <algorithm>
#include <chrono>
#include <thread>

using namespace std;

// Simple matrix operations for neural networks (using vectors)
using Matrix = vector<vector<double>>;
using Vector = vector<double>;

// Utility functions for matrix operations
Matrix matrix_multiply(const Matrix& A, const Matrix& B) {
    int rowsA = A.size();
    int colsA = A[0].size();
    int colsB = B[0].size();
    Matrix C(rowsA, vector<double>(colsB, 0.0));
    for (int i = 0; i < rowsA; ++i) {
        for (int j = 0; j < colsB; ++j) {
            for (int k = 0; k < colsA; ++k) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}

Vector matrix_vector_multiply(const Matrix& A, const Vector& v) {
    int rows = A.size();
    int cols = A[0].size();
    Vector result(rows, 0.0);
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            result[i] += A[i][j] * v[j];
        }
    }
    return result;
}

Matrix matrix_add(const Matrix& A, const Matrix& B) {
    int rows = A.size();
    int cols = A[0].size();
    Matrix C(rows, vector<double>(cols, 0.0));
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            C[i][j] = A[i][j] + B[i][j];
        }
    }
    return C;
}

Vector vector_add(const Vector& a, const Vector& b) {
    int n = a.size();
    Vector c(n);
    for (int i = 0; i < n; ++i) {
        c[i] = a[i] + b[i];
    }
    return c;
}

Vector sigmoid(const Vector& x) {
    int n = x.size();
    Vector result(n);
    for (int i = 0; i < n; ++i) {
        result[i] = 1.0 / (1.0 + exp(-x[i]));
    }
    return result;
}

Vector relu(const Vector& x) {
    int n = x.size();
    Vector result(n);
    for (int i = 0; i < n; ++i) {
        result[i] = max(0.0, x[i]);
    }
    return result;
}

double dot_product(const Vector& a, const Vector& b) {
    double sum = 0.0;
    for (size_t i = 0; i < a.size(); ++i) {
        sum += a[i] * b[i];
    }
    return sum;
}

// Activation function for GNN messages
Vector gnn_aggregate(const vector<Vector>& messages) {
    // Simple mean aggregation
    if (messages.empty()) return {};
    size_t dim = messages[0].size();
    Vector agg(dim, 0.0);
    for (const auto& msg : messages) {
        agg = vector_add(agg, msg);
    }
    for (size_t i = 0; i < dim; ++i) {
        agg[i] /= messages.size();
    }
    return agg;
}

// Simple Graph Neural Network layer
Vector gnn_layer(const Vector& node_features, const vector<Vector>& neighbor_features, const Matrix& W) {
    // Message passing: aggregate neighbor features
    vector<Vector> messages;
    for (const auto& nf : neighbor_features) {
        Matrix nf_mat(1, nf);  // Treat as 1xD matrix
        Matrix msg = matrix_multiply(nf_mat, W);
        messages.push_back(msg[0]);
    }
    Vector agg = gnn_aggregate(messages);
    // Combine with self
    Vector self_msg = matrix_vector_multiply({node_features}, node_features);  // Identity
    Vector combined = vector_add(self_msg, agg);
    // Apply linear transformation
    Matrix W_trans(1, W[0].size());  // Dummy
    return relu(matrix_vector_multiply(W, combined));
}

// Job Shop Scheduling Problem Representation
struct Operation {
    int job_id;
    int machine_id;
    int processing_time;
    bool is_scheduled = false;
    double start_time = 0.0;
    double end_time = 0.0;
};

class JSSP {
public:
    int num_jobs;
    int num_machines;
    vector<vector<Operation>> operations;  // operations[job][op_index]
    vector<double> machine_available;      // available time for each machine
    vector<double> job_ready;              // ready time for next op in job
    double makespan = 0.0;

    JSSP(int nj, int nm) : num_jobs(nj), num_machines(nm),
        machine_available(nm, 0.0), job_ready(nj, 0.0) {
        operations.resize(nj);
    }

    void add_job_operations(int job, const vector<Operation>& ops) {
        operations[job] = ops;
    }

    // Get unscheduled operations that are ready
    vector<int> get_ready_operations() {
        vector<int> ready;
        for (int j = 0; j < num_jobs; ++j) {
            for (int o = 0; o < operations[j].size(); ++o) {
                auto& op = operations[j][o];
                if (!op.is_scheduled && job_ready[j] <= op.start_time) {  // Simplified
                    ready.push_back(j * 100 + o);  // Unique id
                }
            }
        }
        return ready;
    }

    // Schedule an operation
    void schedule_operation(int op_id) {
        int job = op_id / 100;
        int o_idx = op_id % 100;
        auto& op = operations[job][o_idx];
        if (op.is_scheduled) return;

        // Find earliest start time
        double start = max(job_ready[job], machine_available[op.machine_id]);
        op.start_time = start;
        op.end_time = start + op.processing_time;
        machine_available[op.machine_id] = op.end_time;
        job_ready[job] = op.end_time;
        op.is_scheduled = true;
        makespan = max(makespan, op.end_time);
    }

    // Compute current state graph features (simplified)
    // Nodes: unscheduled operations, features: [processing_time, machine_id, job_id, ready_time]
    pair<vector<Vector>, vector<vector<int>>> get_graph_state() {
        vector<int> ready_ops = get_ready_operations();
        int n_nodes = ready_ops.size();
        vector<Vector> node_features(n_nodes, vector<double>(4, 0.0));
        vector<vector<int>> adj_list(n_nodes);  // Simplified adjacency, e.g., no edges for demo

        for (int i = 0; i < n_nodes; ++i) {
            int op_id = ready_ops[i];
            int job = op_id / 100;
            int o_idx = op_id % 100;
            auto& op = operations[job][o_idx];
            node_features[i][0] = op.processing_time;
            node_features[i][1] = op.machine_id;
            node_features[i][2] = job;
            node_features[i][3] = job_ready[job];  // ready time
        }

        // Add some dummy edges for demonstration (e.g., connect to potential conflicts)
        for (int i = 0; i < n_nodes; ++i) {
            for (int j = i + 1; j < n_nodes; ++j) {
                if (node_features[i][1] == node_features[j][1]) {  // Same machine
                    adj_list[i].push_back(j);
                    adj_list[j].push_back(i);
                }
            }
        }

        return {node_features, adj_list};
    }
};

// Simple Policy Network using GNN embeddings
class PolicyNetwork {
private:
    Matrix W_gnn;  // GNN weight matrix, say 4x8
    Matrix W_policy;  // Policy head, 8x action_dim
    double learning_rate = 0.01;

public:
    PolicyNetwork(int feature_dim, int hidden_dim, int action_dim) {
        // Initialize weights randomly
        random_device rd;
        mt19937 gen(rd());
        uniform_real_distribution<> dis(-1.0, 1.0);

        W_gnn = Matrix(feature_dim, vector<double>(hidden_dim));
        for (auto& row : W_gnn) {
            for (auto& val : row) val = dis(gen);
        }

        W_policy = Matrix(hidden_dim, vector<double>(action_dim));
        for (auto& row : W_policy) {
            for (auto& val : row) val = dis(gen);
        }
    }

    // Get action logits from graph state
    Vector get_action_logits(const vector<Vector>& node_features, const vector<vector<int>>& adj_list) {
        int n_nodes = node_features.size();
        vector<Vector> embeddings(n_nodes, vector<double>(W_gnn[0].size(), 0.0));

        // GNN message passing (1 layer for simplicity)
        for (int i = 0; i < n_nodes; ++i) {
            vector<Vector> neighbor_features;
            for (int neigh : adj_list[i]) {
                neighbor_features.push_back(embeddings[neigh]);  // Use current embeddings (simplified, iterative update needed)
            }
            embeddings[i] = gnn_layer(node_features[i], neighbor_features, W_gnn);
        }

        // Global pooling: mean over all nodes
        Vector global_state(W_gnn[0].size(), 0.0);
        for (const auto& emb : embeddings) {
            global_state = vector_add(global_state, emb);
        }
        for (auto& val : global_state) val /= n_nodes;

        // Policy head
        Vector logits = matrix_vector_multiply(W_policy, global_state);
        return logits;
    }

    // Sample action from logits (softmax)
    int sample_action(const Vector& logits) {
        Vector probs(logits.size());
        double sum_exp = 0.0;
        for (double l : logits) {
            double exp_l = exp(l);
            probs.push_back(exp_l);  // Wait, probs = exp(logits)
            sum_exp += exp_l;
        }
        for (auto& p : probs) p /= sum_exp;

        random_device rd;
        mt19937 gen(rd());
        discrete_distribution<> dist(probs.begin(), probs.end());
        return dist(gen);
    }

    // REINFORCE update (simple policy gradient)
    void update(const Vector& logits, int action, double reward, double baseline = 0.0) {
        // Compute log prob of action
        Vector log_probs(logits.size());
        double max_l = *max_element(logits.begin(), logits.end());
        double sum_exp = 0.0;
        for (size_t i = 0; i < logits.size(); ++i) {
            double exp_l = exp(logits[i] - max_l);
            log_probs[i] = log(exp_l);  // Simplified
            sum_exp += exp_l;
        }
        double log_prob_action = log(exp(logits[action] - max_l) / sum_exp);

        double advantage = reward - baseline;
        // Update weights (gradient ascent on log_prob * advantage)
        // Simplified: adjust W_policy
        for (size_t j = 0; j < W_policy.size(); ++j) {
            for (size_t k = 0; k < W_policy[0].size(); ++k) {
                W_policy[j][k] += learning_rate * advantage * log_prob_action;  // Very rough approximation
            }
        }
    }
};

// Reinforcement Learning Agent
class RLAgent {
private:
    PolicyNetwork policy;
    double gamma = 0.99;  // Discount factor

public:
    RLAgent(int feature_dim, int hidden_dim, int max_actions)
        : policy(feature_dim, hidden_dim, max_actions) {}

    // Solve the JSSP using RL
    double solve(JSSP& problem) {
        vector<int> ready_ops = problem.get_ready_operations();
        int total_ops = 0;
        for (const auto& job_ops : problem.operations) total_ops += job_ops.size();

        double total_reward = 0.0;
        double episode_reward = 0.0;

        while (ready_ops.size() > 0) {
            auto [node_features, adj_list] = problem.get_graph_state();
            int n_actions = node_features.size();

            Vector logits = policy.get_action_logits(node_features, adj_list);
            int action = policy.sample_action(logits);  // action is index in ready_ops

            int op_id = ready_ops[action];
            problem.schedule_operation(op_id);

            // Compute immediate reward (negative delay or something)
            double reward = -problem.makespan;  // Simple, penalize makespan increase

            episode_reward += reward;
            total_reward += episode_reward * pow(gamma, ready_ops.size());

            // Update policy (REINFORCE, but batched here for simplicity)
            policy.update(logits, action, reward);

            ready_ops = problem.get_ready_operations();
            this_thread::sleep_for(chrono::milliseconds(1));  // Simulate time
        }

        return -problem.makespan;  // Negative makespan as reward
    }
};

// Example usage
int main() {
    // Create a small JSSP instance: 2 jobs, 2 machines
    JSSP problem(2, 2);

    // Job 0: Op0 on M0 p=3, Op1 on M1 p=2
    problem.add_job_operations(0, { {0, 0, 3}, {0, 1, 2} });
    // Job 1: Op0 on M1 p=2, Op1 on M0 p=4
    problem.add_job_operations(1, { {1, 1, 2}, {1, 0, 4} });

    RLAgent agent(4, 8, 10);  // feature_dim=4, hidden=8, max_actions=10

    double reward = agent.solve(problem);

    cout << "Makespan: " << problem.makespan << endl;
    cout << "Total Reward: " << reward << endl;

    // Print schedule
    for (int j = 0; j < problem.num_jobs; ++j) {
        for (const auto& op : problem.operations[j]) {
            if (op.is_scheduled) {
                cout << "Job " << op.job_id << " Op on M" << op.machine_id
                     << " from " << op.start_time << " to " << op.end_time << endl;
            }
        }
    }

    return 0;
}

File: graph_neural_network/Python/Grok4_JSSP_GNN.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from typing import List, Tuple
from torch.distributions import Categorical

class Operation:
    def __init__(self, job_id: int, machine_id: int, processing_time: int):
        self.job_id = job_id
        self.machine_id = machine_id
        self.processing_time = processing_time
        self.is_scheduled = False
        self.start_time = 0.0
        self.end_time = 0.0

class JSSP:
    def __init__(self, num_jobs: int, num_machines: int):
        self.num_jobs = num_jobs
        self.num_machines = num_machines
        self.operations = [[] for _ in range(num_jobs)]
        self.machine_available = [0.0] * num_machines
        self.job_ready = [0.0] * num_jobs
        self.next_op = [0] * num_jobs
        self.makespan = 0.0

    def add_job(self, job_id: int, ops: List[Operation]):
        self.operations[job_id] = ops

    def get_ready_operations(self) -> List[Tuple[int, int]]:
        ready = []
        for j in range(self.num_jobs):
            if self.next_op[j] < len(self.operations[j]):
                op_idx = self.next_op[j]
                ready.append((j, op_idx))
        return ready

    def schedule_operation(self, job: int, op_idx: int):
        if op_idx != self.next_op[job]:
            return
        op = self.operations[job][op_idx]
        if op.is_scheduled:
            return
        start = max(self.job_ready[job], self.machine_available[op.machine_id])
        op.start_time = start
        op.end_time = start + op.processing_time
        self.machine_available[op.machine_id] = op.end_time
        self.job_ready[job] = op.end_time
        op.is_scheduled = True
        self.next_op[job] += 1
        self.makespan = max(self.makespan, op.end_time)

    def get_graph_state(self) -> Tuple[List[List[float]], List[List[int]]]:
        ready_ops = self.get_ready_operations()
        n_nodes = len(ready_ops)
        node_features = [[0.0] * 4 for _ in range(n_nodes)]
        adj_list = [[] for _ in range(n_nodes)]

        machine_to_nodes = {}
        for i, (j, o_idx) in enumerate(ready_ops):
            op = self.operations[j][o_idx]
            node_features[i] = [float(op.processing_time), float(op.machine_id), float(j), self.job_ready[j]]
            m = op.machine_id
            if m not in machine_to_nodes:
                machine_to_nodes[m] = []
            machine_to_nodes[m].append(i)

        for m_nodes in machine_to_nodes.values():
            for ii in range(len(m_nodes)):
                for jj in range(ii + 1, len(m_nodes)):
                    i1, i2 = m_nodes[ii], m_nodes[jj]
                    adj_list[i1].append(i2)
                    adj_list[i2].append(i1)

        return node_features, adj_list

class GNNLayer(nn.Module):
    def __init__(self, input_dim: int, hidden_dim: int):
        super().__init__()
        self.linear_self = nn.Linear(input_dim, hidden_dim)
        self.linear_neigh = nn.Linear(input_dim, hidden_dim)
        self.linear_combine = nn.Linear(hidden_dim * 2, hidden_dim)

    def forward(self, x: torch.Tensor, adj: torch.Tensor) -> torch.Tensor:
        h_self = F.relu(self.linear_self(x))
        h_neigh_input = F.relu(self.linear_neigh(x))
        h_neigh = torch.mm(adj, h_neigh_input)
        degrees = torch.sum(adj, dim=1, keepdim=True) + 1e-6
        h_neigh = h_neigh / degrees
        h_combined = torch.cat([h_self, h_neigh], dim=1)
        return F.relu(self.linear_combine(h_combined))

class PolicyNetwork(nn.Module):
    def __init__(self, feature_dim: int, hidden_dim: int):
        super().__init__()
        self.gnn = GNNLayer(feature_dim, hidden_dim)
        self.policy_head = nn.Linear(hidden_dim, 1)

    def forward(self, node_features: torch.Tensor, adj: torch.Tensor) -> torch.Tensor:
        embeddings = self.gnn(node_features, adj)
        logits = self.policy_head(embeddings).squeeze(-1)
        return logits

def build_adj_matrix(adj_list: List[List[int]], n_nodes: int) -> torch.Tensor:
    adj = torch.zeros(n_nodes, n_nodes, dtype=torch.float)
    for i, neighs in enumerate(adj_list):
        for j in neighs:
            adj[i, j] = 1.0
    return adj

class RLAgent:
    def __init__(self, feature_dim: int = 4, hidden_dim: int = 8):
        self.policy = PolicyNetwork(feature_dim, hidden_dim)
        self.optimizer = torch.optim.Adam(self.policy.parameters(), lr=0.01)

    def get_action_log_prob(self, node_features: List[List[float]], adj_list: List[List[int]]) -> Tuple[int, torch.Tensor]:
        n_nodes = len(node_features)
        x = torch.FloatTensor(node_features)
        adj = build_adj_matrix(adj_list, n_nodes)
        logits = self.policy(x, adj)
        action_dist = Categorical(logits=logits)
        action = action_dist.sample().item()
        log_prob = action_dist.log_prob(torch.tensor(action))
        return action, log_prob

    def update(self, log_prob: torch.Tensor, reward: float, baseline: float = 0.0):
        advantage = reward - baseline
        loss = -log_prob * advantage
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()

def solve_jssp(agent: RLAgent, problem: JSSP) -> float:
    total_reward = 0.0
    while True:
        ready_ops = problem.get_ready_operations()
        if not ready_ops:
            break
        node_features, adj_list = problem.get_graph_state()
        action, log_prob = agent.get_action_log_prob(node_features, adj_list)
        job, op_idx = ready_ops[action]
        prev_makespan = problem.makespan
        problem.schedule_operation(job, op_idx)
        reward = -(problem.makespan - prev_makespan)
        agent.update(log_prob, reward)
        total_reward += reward
    return total_reward

# Example usage
if __name__ == "__main__":
    problem = JSSP(2, 2)
    problem.add_job(0, [Operation(0, 0, 3), Operation(0, 1, 2)])
    problem.add_job(1, [Operation(1, 1, 2), Operation(1, 0, 4)])
    agent = RLAgent()
    reward = solve_jssp(agent, problem)
    print(f"Makespan: {problem.makespan}")
    print(f"Total Reward: {reward}")
    for j in range(problem.num_jobs):
        for op in problem.operations[j]:
            if op.is_scheduled:
                print(f"Job {op.job_id} on M{op.machine_id} [{op.start_time:.1f}, {op.end_time:.1f}]")